### ISSUE TITLE
Update to latest version of clean-css

### ISSUE THREAD

**Issue by jonsquared** [scniro/gulp-clean-css#31]
Please upgrade to the latest version of clean-css (at this time it is up to version 4.0.0)

**Comment by scniro** [comment]
@jonsquared dang dude, it's been one day since [4.0.0](https://github.com/jakubpawlowicz/clean-css/releases/tag/v4.0.0). I'll get to it soon. It's a pretty big release and I might need to rework some stuff. I assure you - it'll be updated.

**Comment by jonsquared** [comment]
Hah, I didn't notice. I just ran into a defect that is fixed in the newer release.

**Comment by davidnknight** [comment]
I have dropped gulp-clean-css as I also ran into issues that are fixed in the latest version of clean-css, now v4.0.4. I cannot rely on a plugin that doesn't get updated for weeks and causes things to break.

For anyone else facing issues, there is another approach for using CleanCSS without this wrapper plugin in Gulp, see the link below.
https://github.com/jakubpawlowicz/clean-css/issues/342

**Comment by scniro** [comment]
@davidnknight gulp-clean-css hasn't been updated to 4.x yet. FWIW, I have been planning to update it some time this weekend. There were a lot of breaking changes regarding the rebasing that I have to investigate. Also, I wouldn't necessarily generalize 11 days as "weeks." Also, unsure what breaking "things" you're referring to.

**Comment by erwstout** [comment]
I'd just like to chime in and say Sal you're doing a great job. People need
to understand this stuff is developed in free time for the love of the
community, there's no need to get so snippy about something being 11 days
behind a release. If it's that important it gets updated asap for you
David, then maybe think about doing a PR and contributing?

**Comment by scniro** [comment]
Upgraded to 4.x with [gulp-clean-css 2.4.0](https://github.com/scniro/gulp-clean-css/releases/tag/2.4.0) release. Can't guarantee it's working as expected with the new `rebaseTo` has thrown a wrench in what I finally thought to be fixed in prior issues #9 #8 #22 

I'll try to get to it sometime soon, but it's been unusually busy on my end with the day-to-day. Any help is welcomed and well appreciated. On that note, @erwstout thank you very much for the positive feedback on this!

Please keep this thread updated with any findings

**Comment by emjaksa** [comment]
@scniro great work on this package and the update. Can I make a suggestion? When you update to a major version of clean-css you also update your package version to a major version number e.g. 3.0.0. If you have "gulp-clean-css": "^2.2.2" in the package.json it will update to the latest compatible with version and 2.4.0 was not.

**Comment by EAnushan** [comment]
Thanks for all your hard work @scniro. I agree with @emjaksa. Do you think you can tag a new major version when there are breaking changes downstream? I wasn't expecting the `rebaseTo` change, so when I updated my packages, all my fonts and images were broken due to being rebased incorrectly.

**Comment by scniro** [comment]
@emjaksa I certainly will! @EAnushan yea I knew that disaster was coming. I was continuously bashing my head over the last implementation and now must attempt to figure it out again, regarding the rebasing that is. If you have any suggestions I will highly appreciate it.

**Comment by scniro** [comment]
@emjaksa FWIW I just released under [3.0.0](https://github.com/scniro/gulp-clean-css/releases/tag/3.0.0) and 2.4.0 has been deprecated. Hopefully that'll clear up any confusion moving forward

**Comment by EAnushan** [comment]
Regarding the `rebaseTo` changes. Here are the changes I had to make to get everything working again. I'm not sure if it'll help you, but I thought I'd share just in case.

I went from something like:
```
return gulp.src(options.css_dir)
.pipe(plugins.cleanCss({
    target: options.css_min_dir
}))
.pipe(gulp.dest(options.css_min_dir));
```

to:

```
return gulp.src(options.css_dir)
.pipe(plugins.cleanCss({
    rebaseTo: path.relative(options.css_dir, options.css_min_dir)
}))
.pipe(gulp.dest(options.css_min_dir));
```

**Comment by lindevs** [comment]
I can't unterstand why clean-css removes @media, for example:
CSS:
```
@media(min-width:768px) {
    #wrapper {
        padding-left: 220px;
    }

    #wrapper.toggled {
        padding-left: 0;
    }

    #sidebar-wrapper {
        width: 220px;
    }

    #wrapper.toggled #sidebar-wrapper {
        width: 0;
    }
}
```
Using gulp-clean-css:
```
var cssMinify = require('gulp-clean-css');
gulp.task('css_processing', function () {
    return gulp.src(cssFiles)
        .pipe(cssMinify())
        .pipe(gulp.dest(buildFolder + '/' + styleFolder));
});
```
I get this result:
`#wrapper.toggled{padding-left:0}#sidebar-wrapper{width:220px}#wrapper.toggled #sidebar-wrapper{width:0}`

When I using gulp-minify-css:
```
var cssMinify = require('gulp-minify-css');
gulp.task('css_processing', function () {
    return gulp.src(cssFiles)
        .pipe(cssMinify())
        .pipe(gulp.dest(buildFolder + '/' + styleFolder));
});
```
I get this results as expected:
`@media(min-width:768px){#wrapper{padding-left:220px}#wrapper.toggled{padding-left:0}#sidebar-wrapper{width:220px}#wrapper.toggled #sidebar-wrapper{width:0}}`
Why gulp-clean-css removes @media?

**Comment by scniro** [comment]
@lindevs  [https://www.npmjs.com/package/gulp-minify-css](gulp-minify-css) has been deprecated long ago and last I checked used clean-css version 3.3.3 see [here](https://github.com/scniro/gulp-clean-css/commit/438a4faf27f134c30e3e94024a83951c21fdc5cc#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R18)

**Comment by lindevs** [comment]
Hi, can you update "clean-css": "^4.0.4" to "clean-css": "^4.0.5"?

**Comment by scniro** [comment]
@lindevs you still did not comment back on my initial reply. I am not understanding the issues you are having, so I can not help ATM without a better explanation and expected result. Spanning an update request on this issue thread with every clean-css version increment the moment it comes out will not likely help you either nor is really appropriate. I will eventually update it.

@EAnushan thank you for sharing your code. Your workaround may be helpful once I dig into this again.

**Comment by scniro** [comment]
Build broke, see fix in 3.0.2 37997b0

**Comment by lindevs** [comment]
@scniro 4.0.5 clean-css update fixed my issue. I tested with [https://jakubpawlowicz.github.io/clean-css/](https://jakubpawlowicz.github.io/clean-css/)

**Comment by wlachenal** [comment]
After updating to newer version, which I'm using @import in css. All @imports are showing errors.

Ignoring local @import of "base/reset.css" as resource is missing.
Ignoring local @import of "base/css-properties.css" as resource is missing.
Ignoring local @import of "fonts/font-styles.css" as resource is missing.
And so on..

Any solution for this stuff?

I already added the inline local thing, but it doesn't work.

```
.pipe(cleanCSS({ keepSpecialComments: 0, inline: 'local' }, function(details) {
    console.log(details.name + ': ' + details.stats.originalSize);
    console.log(details.name + ': ' + details.stats.minifiedSize);
}))
```

**Comment by corneyl** [comment]
@wlachenal @scniro Imports seem to fail because `CleanCSS` doesn't know the source folder of the css file, only the source file content is passed to the `CleanCSS().minify()` function. So it looks like the imports are being imported from the CWD. When I run `CleanCSS` directly, imports are working fine...

**Comment by corneyl** [comment]
@scniro any comments on my proposed solution? The newest published version doesn't work for me and probably a lot of other people.

**Comment by scniro** [comment]
@corneyl I apologize for the delay. I can in the least, commit to looking into this extensively this weekend when I have the time. Until then, when you mention breaking tests on your PR, could you explain just a bit more? The [CI log](https://travis-ci.org/scniro/gulp-clean-css/builds/209528923) is sating all checks have passes but perhaps I am not seeing all that your are?

**Comment by corneyl** [comment]
I just ran `npm test` and saw some errors, but that was probably due to my local environment, can't remember exactly.

**Comment by scniro** [comment]
@corneyl I may need to investigate this more. The changes in your PR break basic functionality with minifying simple styles. At this time I'd recommend looking into the proposed workaround on #37 supplementing the [gulp-url-rebase](https://github.com/42Zavattas/gulp-css-url-rebase) plugin, or reverting to older plugins which leverage clean-css 3.x.

One of the trade-offs with the 4.x clean-css life-cycle was the favoritism movement for a  simpler rebasing strategy over robust configuration on the previous _three_ parameters (no consolidated to `rebaseTo`

> root, relativeTo, and target options are replaced by a single rebaseTo option - this means that rebasing URLs and import inlining is much simpler but may not be (YMMV) as powerful as in 3.x;

You suggested that this may affect a lot of people which I don't feel is the case or I'd have a lot more buzz on this topic and can hopefully get the help I need on this to solve it.

If you can provide a small, isolated real world fork-able project which shows this exact problem with defined expectations, I'll happily take a look to find the solution. Otherwise I'm kind of shooting in the dark here. I'm sorry I can't be more helpful at this time.

**Comment by corneyl** [comment]
Sorry for the delay, but I've created a simple example as requested: https://github.com/corneyl/gulp-clean-css-example
The `use-filename` branch uses the proposed fix, and `master` uses the latest version (`3.0.4`).

**Comment by scniro** [comment]
@corneyl I too am sorry for the delay. I will clone your sample repo here and work this out this weekend now that I have an isolated example. I've drawn a lot of attention to the issue at hand so It'll get resolved one way or another. Will keep you posted and thanks again for the sample

**Comment by scniro** [comment]
I have revisited this and hope the newest changes will resolve this. Please pull down [3.1.0](https://github.com/scniro/gulp-clean-css/releases/tag/3.1.0) and let me know if this meets your expectations. Thanks!

**Comment by scniro** [comment]
Did some more work on this as seen at jakubpawlowicz/clean-css#937 and all should be fixed with the [3.2.0 release](https://github.com/scniro/gulp-clean-css/releases/tag/3.3.0). I am going to close this for now, but please reply here or open up a new issue should you still encounter issues. Thank you!
